File: examples/calibration/evolution/sources/marble.ts
type InputRecord = { id: string; value: string };
type RecordView = { id: string; value: string };
type Receipt = { size: number };

class Writer {
  published: RecordView[] | null = null;

  publish(batch: RecordView[]): Receipt {
    this.published = batch.slice();
    return { size: batch.length };
  }
}

function toView(record: InputRecord): RecordView {
  return { id: record.id, value: record.value };
}

export function publishBatch(writer: Writer, records: InputRecord[]): Receipt {
  const batch: RecordView[] = [];
  for (const record of records) {
    batch.push(toView(record));
  }
  return writer.publish(batch);
}

File: examples/calibration/evolution/sources/marble.rs
struct Record {
    id: String,
    value: String,
}

struct RecordView {
    id: String,
    value: String,
}

struct Receipt {
    size: usize,
}

struct Writer {
    published: Option<Vec<RecordView>>,
}

impl Writer {
    fn publish(&mut self, batch: Vec<RecordView>) -> Receipt {
        let size = batch.len();
        self.published = Some(batch);
        Receipt { size }
    }
}

fn to_view(record: Record) -> RecordView {
    RecordView {
        id: record.id,
        value: record.value,
    }
}

fn publish_batch(writer: &mut Writer, records: Vec<Record>) -> Receipt {
    let mut batch = Vec::new();
    for record in records {
        batch.push(to_view(record));
    }
    writer.publish(batch)
}
